File: server/routes/audio.py
import json
import math

from flask import Blueprint, jsonify, request

from db import get_db
from limiter import limiter
from utils.youtube_utils import extract_video_id, download_audio_pcm, fetch_video_metadata
from utils.audio_analyzer import analyze_audio

audio_bp = Blueprint('audio', __name__)
ANALYSIS_VERSION = 2



def _seed_for_video(video_id):
    seed = 0
    for char in f'{video_id}:{ANALYSIS_VERSION}':
        seed = ((seed * 131) + ord(char)) & 0xFFFFFFFF
    return seed



def synthetic_analysis(video_id, duration=210):
    """Deterministic fallback used when server-side extraction is unavailable."""
    seed = _seed_for_video(video_id)
    bpm = 92 + (seed % 74)
    beat_interval = 60 / bpm

    beat_timestamps = []
    t = 0.0
    while t < duration:
        beat_timestamps.append(round(t, 3))
        t += beat_interval

    waveform = []
    for i in range(96):
        waveform.append(round(0.35 + 0.3 * math.sin((i / 96) * math.pi * 4 + (seed % 11)), 4))

    frequency_frames = []
    t = 0.0
    while t < duration:
        bass = 0.5 + 0.4 * math.sin(t * 1.4 + (seed % 5))
        mid = 0.5 + 0.35 * math.sin(t * 2.2 + (seed % 7))
        high = 0.5 + 0.3 * math.sin(t * 3.8 + (seed % 13))
        frame = {
            't': round(t, 3),
            'sub': round(max(0.0, min(1.0, bass * 0.9)), 3),
            'bass': round(max(0.0, min(1.0, bass)), 3),
            'lowMid': round(max(0.0, min(1.0, (bass * 0.45 + mid * 0.55))), 3),
            'mid': round(max(0.0, min(1.0, mid)), 3),
            'highMid': round(max(0.0, min(1.0, (mid * 0.5 + high * 0.5))), 3),
            'presence': round(max(0.0, min(1.0, high * 0.9)), 3),
            'brilliance': round(max(0.0, min(1.0, high * 0.8)), 3),
            'air': round(max(0.0, min(1.0, high * 0.6)), 3),
            'onset': round(max(0.0, min(1.0, abs(math.sin(t * 3.2)))), 3),
            'beatConfidence': round(max(0.0, min(1.0, 1 - abs((t % beat_interval) - beat_interval * 0.5) / (beat_interval * 0.5))), 3),
            'pan': round(math.sin(t * 0.9 + (seed % 9)) * 0.7, 3),
            'centroid': round(max(0.0, min(1.0, 0.35 + high * 0.45)), 3),
            'events': [],
        }
        if frame['bass'] > 0.72 and frame['onset'] > 0.42:
            frame['events'].append('kick')
        if frame['mid'] > 0.62 and frame['onset'] > 0.34:
            frame['events'].append('snare')
        if frame['presence'] > 0.68:
            frame['events'].append('hat')
        if frame['highMid'] > 0.58:
            frame['events'].append('vocal')
        frequency_frames.append(frame)
        t += 0.1

    return {
        'bpm': bpm,
        'beatTimestamps': beat_timestamps,
        'frequencyFrames': frequency_frames,
        'waveform': waveform,
    }



def _cached_payload(row):
    payload = json.loads(row['analysis_json'])
    return {
        'videoId': row['video_id'],
        'title': payload.get('title') or row['title'],
        'artist': payload.get('artist', 'YouTube'),
        'thumbnail': payload.get('thumbnail'),
        'bpm': row['bpm'],
        'duration': payload.get('duration', row['duration']),
        **payload,
    }


@audio_bp.get('/analyze')
@limiter.limit('20 per minute')
def analyze():
    """
    GET /api/audio/analyze?url=<youtubeUrl>
    Returns cached or freshly computed analysis for a YouTube video.
    """
    url = request.args.get('url')
    if not url:
        return jsonify({'error': 'url query param required'}), 400

    video_id = extract_video_id(url)
    if not video_id:
        return jsonify({'error': 'Unsupported YouTube or YouTube Music link'}), 400

    conn = get_db()
    cached = conn.execute('SELECT * FROM tracks WHERE video_id = ?', (video_id,)).fetchone()
    if cached:
        payload = json.loads(cached['analysis_json'])
        if payload.get('analysisVersion') == ANALYSIS_VERSION:
            return jsonify(_cached_payload(cached))

    try:
        metadata = fetch_video_metadata(video_id)
        seed = _seed_for_video(video_id)
        duration = 0
        try:
            pcm, sample_rate = download_audio_pcm(video_id)
            duration = pcm.shape[0] / sample_rate
            analysis = analyze_audio(pcm, sample_rate)
        except Exception as download_error:
            print(f'yt-dlp unavailable ({download_error}) — using seeded synthetic analysis')
            analysis = synthetic_analysis(video_id)
            duration = analysis.get('duration', 210)

        payload = {
            **analysis,
            'analysisVersion': ANALYSIS_VERSION,
            'analysisKey': f'{video_id}:v{ANALYSIS_VERSION}',
            'seed': seed,
            'title': metadata['title'],
            'artist': metadata['artist'],
            'thumbnail': metadata['thumbnail'],
            'duration': duration,
        }

        conn.execute(
            'INSERT OR REPLACE INTO tracks (video_id, title, duration, bpm, analysis_json) VALUES (?, ?, ?, ?, ?)',
            (video_id, metadata['title'], duration, analysis['bpm'], json.dumps(payload))
        )
        conn.commit()

        return jsonify({'videoId': video_id, 'bpm': analysis['bpm'], **payload})
    except Exception as err:
        print(f'Audio analysis error: {err}')
        return jsonify({'error': 'Analysis unavailable right now'}), 500


@audio_bp.get('/status/<video_id>')
@limiter.limit('120 per minute')
def status(video_id):
    """GET /api/audio/status/<videoId> — quick check if video is already analysed."""
    conn = get_db()
    row = conn.execute(
        'SELECT video_id, title, bpm, duration, analysis_json FROM tracks WHERE video_id = ?', (video_id,)
    ).fetchone()
    if row:
        payload = json.loads(row['analysis_json']) if row.get('analysis_json') else {}
        return jsonify({
            'analyzed': payload.get('analysisVersion') == ANALYSIS_VERSION,
            'videoId': row['video_id'],
            'title': payload.get('title') or row['title'],
            'artist': payload.get('artist', 'YouTube'),
            'thumbnail': payload.get('thumbnail'),
            'bpm': row['bpm'],
            'duration': payload.get('duration', row['duration']),
            'analysisKey': payload.get('analysisKey'),
            'analysisVersion': payload.get('analysisVersion'),
            'seed': payload.get('seed'),
        })
    return jsonify({'analyzed': False})

File: server/routes/audio.js
'use strict';

const express = require('express');
const router = express.Router();
const { getDb } = require('../db');
const { extractVideoId, downloadAudioPCM, fetchVideoTitle } = require('../utils/youtubeUtils');
const { analyzeAudio } = require('../utils/audioAnalyzer');

/**
 * GET /api/audio/analyze?url=<youtubeUrl>
 *
 * Returns cached or freshly computed analysis for a YouTube video.
 * Analysis includes: bpm, beatTimestamps[], frequencyFrames[], title, duration.
 *
 * If yt-dlp is unavailable, returns a synthetic analysis seeded from the
 * video ID so the game is still playable.
 */
router.get('/analyze', async (req, res) => {
  const { url } = req.query;
  if (!url) return res.status(400).json({ error: 'url query param required' });

  const videoId = extractVideoId(url);
  if (!videoId) return res.status(400).json({ error: 'Could not extract video ID from URL' });

  const db = getDb();

  // Return cached analysis if available
  const cached = db.prepare('SELECT * FROM tracks WHERE video_id = ?').get(videoId);
  if (cached) {
    return res.json({
      videoId: cached.video_id,
      title: cached.title,
      bpm: cached.bpm,
      duration: cached.duration,
      ...JSON.parse(cached.analysis_json)
    });
  }

  // Trigger analysis (may take up to ~2 min on large tracks)
  try {
    const title = await fetchVideoTitle(videoId);
    let analysis;
    let duration = 0;

    try {
      const { pcm, sampleRate } = await downloadAudioPCM(videoId);
      duration = pcm.length / sampleRate;
      analysis = analyzeAudio(pcm, sampleRate);
    } catch (dlErr) {
      console.warn(`yt-dlp unavailable (${dlErr.message}) — using seeded synthetic analysis`);
      analysis = syntheticAnalysis(videoId);
    }

    // Persist
    db.prepare(`
      INSERT OR REPLACE INTO tracks (video_id, title, duration, bpm, analysis_json)
      VALUES (?, ?, ?, ?, ?)
    `).run(videoId, title, duration, analysis.bpm, JSON.stringify({
      beatTimestamps: analysis.beatTimestamps,
      frequencyFrames: analysis.frequencyFrames
    }));

    return res.json({ videoId, title, bpm: analysis.bpm, duration, ...analysis });

  } catch (err) {
    console.error('Audio analysis error:', err);
    return res.status(500).json({ error: 'Analysis failed', detail: err.message });
  }
});

/**
 * GET /api/audio/status/:videoId
 * Quick check — is this video already analysed?
 */
router.get('/status/:videoId', (req, res) => {
  const db = getDb();
  const row = db.prepare('SELECT video_id, title, bpm, duration FROM tracks WHERE video_id = ?')
    .get(req.params.videoId);
  if (row) return res.json({ analyzed: true, ...row });
  return res.json({ analyzed: false });
});

// ─────────────────────────────────────────────────────────────────────────────
// Synthetic analysis: deterministic BPM + beats seeded from videoId
// Used as fallback when yt-dlp is not installed.
// ─────────────────────────────────────────────────────────────────────────────
function syntheticAnalysis(videoId) {
  // Hash videoId to a stable BPM in [80, 180]
  let hash = 0;
  for (const c of videoId) hash = (hash * 31 + c.charCodeAt(0)) >>> 0;
  const bpm = 80 + (hash % 101); // 80–180
  const beatInterval = 60 / bpm;
  // Synthetic analysis assumes a 210-second (3.5 min) track
  const duration = 210;
  const beatTimestamps = [];
  for (let t = 0; t < duration; t += beatInterval) {
    beatTimestamps.push(parseFloat(t.toFixed(3)));
  }

  // Simple sine-wave frequency frames
  const frequencyFrames = [];
  for (let t = 0; t < duration; t += 0.1) {
    frequencyFrames.push({
      t: parseFloat(t.toFixed(1)),
      bass: parseFloat((0.5 + 0.5 * Math.sin((t / beatInterval) * Math.PI)).toFixed(3)),
      mid: parseFloat((0.5 + 0.5 * Math.sin((t / beatInterval) * Math.PI * 2 + 1)).toFixed(3)),
      high: parseFloat((0.5 + 0.5 * Math.sin((t / beatInterval) * Math.PI * 3 + 2)).toFixed(3))
    });
  }

  return { bpm, beatTimestamps, frequencyFrames };
}

module.exports = router;
